feat: hmr docs (#77)
adonisjs/v6-docs: content/docs/fundamentals/hmr.md

# Hot Module Reloading

> Hot Module Reloading (HMR) is a feature baked in since AdonisJS 6.x.x that allows automatic reloading of modified files without having to restart the entire server.

Author: update

Until recently, the traditional way of working with AdonisJS was to launch the dev server with the command `node ace serve --watch`, which would take care of completely restarting the server/node process each time a file was modified, just like Nodemon does for example.

This process works well, but depending on certain things, it's sometimes not the most optimal:

- Sometimes you have a lot of configured packages, and therefore providers that execute on each restart. This can slow down the startup process.

> - If you use TSX as a template engine, then a simple modification of, let's say, paddings or margins in one of your views forces you to reload the entire server, resulting in a quite slow feedback loop.

Reviewer: This is an example linked to using a utility-first CSS framework. I don't believe this example should be shared here since it relates to other technology and not directly in TSX. We should explain that TSX views are JavaScript modules and therefore, are cached the same way as a JavaScript file.

> To correct these issues and improve the development experience, AdonisJS 6.x.x introduced Hot Module Reloading (HMR) based on [Hot-Hook](https://github.com/julien-R44/hot-hook), also developed by the Core Team.

Author: update version

:::tip

Make sure to read the extensive documentation of [Hot-Hook](https://github.com/julien-R44/hot-hook) to understand in more detail how it works.

:::

## Configuration

> First, make sure you have AdonisJS version 6.x.x. Then, you can configure HMR in your `package.json`. You will generally only need to specify the list of your controllers via a glob pattern.

Author: update version

```jsonc
// title: package.json
{
  "name": "adonis-app",
  "type": "module",
  // ...
  "hotHook": {
    "boundaries": [
      "./app/controllers/**/*.ts"
    ]
  }
}
```

:::tip
What is a boundary? Make sure to read the [Hot-Hook documentation](https://github.com/Julien-R44/hot-hook#boundary) about this concept. 

> TL;DR: In Hot Hook, modules marked as "boundaries" create an area where HMR is applied, allowing for dynamic reloading of these modules and their dependencies as long as the dependency path to the entrypoint file includes at least one boundary-marked module. 

Reviewer: > at least one boundary-marked module. 

That's not true; all paths to the root should go through a boundary-marked module to allow hot reloading. Otherwise, we are doing a cold reload.

Author: we are not talking about multiple paths here. Just one : 

> as long as the dependency path to the entrypoint file

i deliberately omitted the explanation when there are several paths to root. otherwise it's too complex. here it's a just quick TLDR

Modifications outside these boundaries require a full server restart. 
:::

## Usage

To launch the server with HMR, you will need to use the command `node ace serve --hmr` instead of `node ace serve --watch`.

```sh
node ace serve --hmr
```

> And there you go. Try modifying a controller and you will see that the server does not restart, and if you execute a request, the modifications will be taken into account: you will always have the latest version of your code.

Reviewer: ```suggestion
If you modify a controller, you will see that the server does not restart, but the modifications will be taken into account: you will always have the latest version of your code.
```

The second solution, sometimes simpler, is to disable HMR for this file using `import.meta.hot?.decline`:

```ts
import.meta.hot?.decline()

export default class UsersController {
  store() {
    setInterval(() => {
      console.log('Hello')
    }, 1000)
  }
}
```

By using `import.meta.hot?.decline`, HMR will not be applied to this file : the server will restart when this file is updated.

## Limitations

### Unreloaded files

In AdonisJS, only your controllers and their dependencies are automatically reloaded. If you modify, for example, a provider or a configuration file, your server will be completely restarted: which makes sense because these files are executed only at the startup of your application. So it is necessary to restart the server for the modifications to be taken into account.

### Dynamic imports

You will need to use dynamic imports for your controllers for HMR to work correctly:

```ts
const UserController = () => import('#controllers/users_controller.js')
router.get('/users', [UserController, 'index'])
```

> If you use the [Eslint configuration](./tooling_config.md#eslint-config) of AdonisJS, this will be automatically done for you.

Reviewer: ```suggestion
If you use the [ESLint configuration](./tooling_config.md#eslint-config) of AdonisJS, the dynamic import of the controller will be automatically applied.
```

  // ...
  "hotHook": {
    "boundaries": [
      "./app/controllers/**/*.ts",

>       "./middlewares/*.ts"

Reviewer 2: Should be "./app/middleware/*.ts"
